tools/buildsys/src/gomod.rs

```diff
@@ -0,0 +1,213 @@
+/*!
+Packages using the Go programming language may have upstream tar archives that
+include only the source code of the project, but not the source code of any
+dependencies. The Go programming language promotes the use of "modules" for
+dependencies. Projects adopting modules will provide `go.mod` and `go.sum` files.
+
+This Rust module extends the functionality of `packages.metadata.build-package.external-files`
+and provides the ability to retrieve and validate dependencies
+declared using Go modules given a tar archive containing a `go.mod` and `go.sum`.
+
+The location where dependencies are retrieved from are controlled by the
+standard environment variables employed by the Go tool: `GOPROXY`, `GOSUMDB`, and
+`GOPRIVATE`. These variables are automatically retrieved from the host environment
+when the docker-go script is invoked.
+
+ */
+
+pub(crate) mod error;
+use error::Result;
+
+use super::manifest;
+use duct::cmd;
+use snafu::{ensure, OptionExt, ResultExt};
+use std::io::Write;
+use std::os::unix::fs::PermissionsExt;
+use std::path::{Path, PathBuf};
+use std::process::Output;
+use std::{env, fs};
+
+pub(crate) struct GoMod;
+
+const GO_MOD_DOCKER_SCRIPT_NAME: &str = "docker-go-script.sh";
+
+// The following bash script is intended to be run within a container
+// using the docker-go tool found in this codebase under tools/docker-go.
+//
+// This script inspects the top level directory found in the upstream archive
+// and uses that as the default Go module path
+// if no explicit module path was provided.
+// It will then untar the archive, vendor the Go dependencies,
+// create a new archive using the {module-path}/vendor directory
+// and name it the output path provided.
+// If no output path was given, it defaults to "bundled-{tar-file-name}".
+// Finally, it cleans up by removing the untar'd source code.
+// The upstream archive remains intact and both tar files can be used during packaging.
+//
+// This script exists as an in memory string literal
+// and is populated into a temporary file in the package directory itself
+// to enable buildsys to be as portable as possible and have no dependecy on runtime paths.
+// Since buildsys is executed from the context of many different package directories,
+// managing a temporary file via this Rust module
+// prevents having to aquire the path of some static script file on the host system.
+const GO_MOD_SCRIPT: &str = r###"#!/bin/bash
+
+set -e
+
+toplevel=$(tar tf "${LOCAL_FILE_NAME}" | head -1)
+if [ -z "${MOD_DIR}" ] ; then
+    targetdir="${toplevel}"
+else
+    targetdir="${MOD_DIR}"
+fi
+
+tar xf "${LOCAL_FILE_NAME}"
+
+pushd "${targetdir}"
+	go list -mod=readonly ./... >/dev/null && go mod vendor
```

Comment:
nit: a tab snuck in here
```suggestion
    go list -mod=readonly ./... >/dev/null && go mod vendor
```